File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/ConnectivityListener.java
package com.amazon.tv.leanbacklauncher.apps;

import android.annotation.TargetApi;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.wifi.WifiConfiguration;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.telephony.PhoneStateListener;
import android.telephony.SignalStrength;
import android.telephony.TelephonyManager;
import android.util.Log;
import java.lang.ref.WeakReference;
import java.util.List;

public class ConnectivityListener {
    private final ConnectivityManager mConnectivityManager;
    private ConnectivityStatus mConnectivityStatus = new ConnectivityStatus();
    private final Context mContext;
    private final IntentFilter mFilter;
    private boolean mIsRegistered;
    private final Listener mListener;
    private final LeanbackLauncherPhoneStateListener mPhoneStateListener;
    private final BroadcastReceiver mReceiver;
    private final TelephonyManager mTelephonyManager;
    private final WifiManager mWifiManager;

    public interface Listener {
        void onConnectivityChange();
    }

    public static class ConnectivityStatus {
        public String mMobileNetworkName;
        public int mMobileSignalStrength;
        public int mNetworkType;
        public int mWifiSignalStrength;
        public String mWifiSsid;
    }

    private static class LeanbackLauncherPhoneStateListener extends PhoneStateListener {
        private WeakReference<ConnectivityListener> mListener;

        public LeanbackLauncherPhoneStateListener(ConnectivityListener listener) {
            this.mListener = new WeakReference(listener);
        }

        public void onSignalStrengthsChanged(SignalStrength signalStrength) {
            super.onSignalStrengthsChanged(signalStrength);
            ConnectivityListener listener = (ConnectivityListener) this.mListener.get();
            if (listener != null) {
                listener.mConnectivityStatus.mMobileSignalStrength = ConnectivityListener.getLevel(signalStrength);
            }
        }
    }

    @TargetApi(23)
    private static int getLevel(SignalStrength signalStrength) {
        LinkageError e;
        try {
            return signalStrength.getLevel();
        } catch (NoClassDefFoundError e2) {
            e = e2;
            Log.e("ConnectivityListener", "Exception fetching signal level", e);
            return 0;
        } catch (IncompatibleClassChangeError e3) {
            e = e3;
            Log.e("ConnectivityListener", "Exception fetching signal level", e);
            return 0;
        }
    }

    public ConnectivityListener(Context context, Listener listener) {
        this.mContext = context;
        this.mConnectivityManager = (ConnectivityManager) this.mContext.getApplicationContext().getSystemService("connectivity");
        this.mWifiManager = (WifiManager) this.mContext.getSystemService("wifi");
        this.mPhoneStateListener = new LeanbackLauncherPhoneStateListener(this);
        this.mTelephonyManager = (TelephonyManager) this.mContext.getSystemService("phone");
        this.mListener = listener;
        this.mFilter = new IntentFilter();
        this.mFilter.addAction("android.net.conn.CONNECTIVITY_CHANGE");
        this.mFilter.addAction("android.net.wifi.RSSI_CHANGED");
        this.mFilter.addAction("android.net.conn.INET_CONDITION_ACTION");
        this.mReceiver = new BroadcastReceiver() {
            public void onReceive(Context context, Intent intent) {
                boolean z = false;
                String intentAction = intent.getAction();
                int connectionStatus = intent.getIntExtra("inetCondition", -551);
                NetworkInfo info = ((ConnectivityManager) context.getSystemService("connectivity")).getActiveNetworkInfo();
                if (!(info != null && info.isAvailable() && info.isConnected())) {
                    ConnectivityListener.writeConnectivity(context, false);
                }
                if (intentAction.equals("android.net.conn.INET_CONDITION_ACTION") || (intentAction.equals("android.net.conn.CONNECTIVITY_CHANGE") && !ConnectivityListener.readConnectivity(ConnectivityListener.this.mContext))) {
                    if (connectionStatus > 50) {
                        z = true;
                    }
                    ConnectivityListener.writeConnectivity(context, z);
                }
                ConnectivityListener.this.updateConnectivityStatus();
                ConnectivityListener.this.mListener.onConnectivityChange();
            }
        };
    }

    private static void writeConnectivity(Context context, boolean inetConnected) {
        context.getSharedPreferences("inet-prefs", 0).edit().putBoolean("inetCondition", inetConnected).apply();
    }

    public static boolean readConnectivity(Context context) {
        return context.getSharedPreferences("inet-prefs", 0).getBoolean("inetCondition", true);
    }

    public void start() {
        if (!this.mIsRegistered) {
            updateConnectivityStatus();
            this.mContext.registerReceiver(this.mReceiver, this.mFilter);
            this.mTelephonyManager.listen(this.mPhoneStateListener, 256);
            this.mIsRegistered = true;
        }
    }

    public void stop() {
        if (this.mIsRegistered) {
            this.mContext.unregisterReceiver(this.mReceiver);
            this.mTelephonyManager.listen(this.mPhoneStateListener, 0);
            this.mIsRegistered = false;
        }
    }

    public ConnectivityStatus getConnectivityStatus() {
        return this.mConnectivityStatus;
    }

    private void setNoConnection() {
        this.mConnectivityStatus.mNetworkType = 1;
        this.mConnectivityStatus.mWifiSsid = null;
        this.mConnectivityStatus.mWifiSignalStrength = 0;
    }

    private boolean isSecureWifi(WifiInfo wifiInfo) {
        if (wifiInfo == null) {
            return false;
        }
        int networkId = wifiInfo.getNetworkId();
        List<WifiConfiguration> configuredNetworks = this.mWifiManager.getConfiguredNetworks();
        if (configuredNetworks == null) {
            return false;
        }
        for (WifiConfiguration configuredNetwork : configuredNetworks) {
            if (configuredNetwork.networkId == networkId) {
                if (configuredNetwork.allowedKeyManagement.get(1) || configuredNetwork.allowedKeyManagement.get(2) || configuredNetwork.allowedKeyManagement.get(3)) {
                    return true;
                }
                return false;
            }
        }
        return false;
    }

    private void updateConnectivityStatus() {
    	// deprecated in API29
        NetworkInfo networkInfo = this.mConnectivityManager.getActiveNetworkInfo();
        if (networkInfo == null) {
            setNoConnection();
            return;
        }
        boolean isConnected = readConnectivity(this.mContext);
        switch (networkInfo.getType()) {
            case 0:
                if (isConnected) {
                    this.mConnectivityStatus.mNetworkType = 13;
                } else {
                    this.mConnectivityStatus.mNetworkType = 15;
                }
                String operator = null;
                if (this.mTelephonyManager != null) {
                    operator = this.mTelephonyManager.getNetworkOperatorName();
                    if (operator != null) {
                        operator = removeDoubleQuotes(operator);
                    }
                }
                if (operator != null) {
                    this.mConnectivityStatus.mMobileNetworkName = operator;
                } else {
                    this.mConnectivityStatus.mMobileNetworkName = "";
                }
                this.mConnectivityStatus.mWifiSsid = null;
                this.mConnectivityStatus.mWifiSignalStrength = 0;
                return;
            case 1:
                WifiInfo wifiInfo = this.mWifiManager.getConnectionInfo();
                if (isSecureWifi(wifiInfo)) {
                    this.mConnectivityStatus.mNetworkType = 5;
                } else {
                    this.mConnectivityStatus.mNetworkType = 3;
                }
                if (!isConnected) {
                    this.mConnectivityStatus.mNetworkType = 9;
                }
                String ssid = null;
                if (wifiInfo != null) {
                    ssid = wifiInfo.getSSID();
                    if (ssid != null) {
                        ssid = removeDoubleQuotes(ssid);
                    }
                }
                if (ssid != null) {
                    this.mConnectivityStatus.mWifiSsid = ssid.equals("<unknown ssid>") ? "" : ssid;
                } else {
                    this.mConnectivityStatus.mWifiSsid = "";
                }
                if (wifiInfo != null) {
                    this.mConnectivityStatus.mWifiSignalStrength = WifiManager.calculateSignalLevel(wifiInfo.getRssi(), 5);
                } else {
                    this.mConnectivityStatus.mWifiSignalStrength = 0;
                }
                this.mConnectivityStatus.mMobileNetworkName = null;
                this.mConnectivityStatus.mMobileSignalStrength = 0;
                return;
            case 9:
                if (isConnected) {
                    this.mConnectivityStatus.mNetworkType = 7;
                } else {
                    this.mConnectivityStatus.mNetworkType = 11;
                }
                this.mConnectivityStatus.mWifiSsid = null;
                this.mConnectivityStatus.mWifiSignalStrength = 0;
                this.mConnectivityStatus.mMobileNetworkName = null;
                this.mConnectivityStatus.mMobileSignalStrength = 0;
                return;
            default:
                setNoConnection();
                return;
        }
    }

    public static String removeDoubleQuotes(String string) {
        if (string == null) {
            return null;
        }
        int length = string.length();
        if (length > 1 && string.charAt(0) == '\"' && string.charAt(length - 1) == '\"') {
            return string.substring(1, length - 1);
        }
        return string;
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/ConnectivityListener.kt
package com.amazon.tv.leanbacklauncher.apps

import android.Manifest
import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.net.wifi.WifiInfo
import android.net.wifi.WifiManager
import android.telephony.PhoneStateListener
import android.telephony.SignalStrength
import android.telephony.TelephonyManager
import android.util.Log
import androidx.core.app.ActivityCompat
import com.amazon.tv.leanbacklauncher.BuildConfig
import java.lang.ref.WeakReference

class ConnectivityListener(private val mContext: Context, listener: Listener) {
    private val mConnectivityManager: ConnectivityManager =
        mContext.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val connectivityStatus = ConnectivityStatus()
    private val mFilter: IntentFilter
    private var mIsRegistered = false
    private val mListener: Listener
    private val mPhoneStateListener: LeanbackLauncherPhoneStateListener =
        LeanbackLauncherPhoneStateListener(this)
    private val mReceiver: BroadcastReceiver
    private val mTelephonyManager: TelephonyManager?
    private val mWifiManager: WifiManager =
        mContext.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager

    interface Listener {
        fun onConnectivityChange()
    }

    init {
        mTelephonyManager = mContext.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
        mListener = listener
        mFilter = IntentFilter()
        mFilter.addAction("android.net.conn.CONNECTIVITY_CHANGE")
        mFilter.addAction("android.net.wifi.RSSI_CHANGED")
        mFilter.addAction("android.net.conn.INET_CONDITION_ACTION")
        mReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                var z = false
                val intentAction = intent.action
                val connectionStatus = intent.getIntExtra("inetCondition", -551)
                val info =
                    (context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager).activeNetworkInfo
                if (!(info != null && info.isAvailable && info.isConnected)) {
                    writeConnectivity(context, false)
                }
                if (intentAction == "android.net.conn.INET_CONDITION_ACTION" || intentAction == "android.net.conn.CONNECTIVITY_CHANGE" && !readConnectivity(
                        mContext
                    )
                ) {
                    if (connectionStatus > 50) {
                        z = true
                    }
                    writeConnectivity(context, z)
                }
                updateConnectivityStatus()
                mListener.onConnectivityChange()
            }
        }
    }

    companion object {
        private val TAG =
            if (BuildConfig.DEBUG) ("*" + javaClass.simpleName).take(21) else javaClass.simpleName

        @TargetApi(23)
        private fun getLevel(signalStrength: SignalStrength): Int {
            val e: LinkageError
            return try {
                signalStrength.level
            } catch (e2: NoClassDefFoundError) {
                e = e2
                Log.e(TAG, "Exception fetching signal level", e)
                0
            } catch (e3: IncompatibleClassChangeError) {
                e = e3
                Log.e(TAG, "Exception fetching signal level", e)
                0
            }
        }

        private fun writeConnectivity(context: Context, inetConnected: Boolean) {
            context.getSharedPreferences("inet-prefs", 0).edit()
                .putBoolean("inetCondition", inetConnected).apply()
        }

        @JvmStatic
        fun readConnectivity(context: Context): Boolean {
            return context.getSharedPreferences("inet-prefs", 0).getBoolean("inetCondition", true)
        }

        fun removeDoubleQuotes(string: String?): String? {
            if (string == null) {
                return null
            }
            return string.removeSurrounding("\"")
//            val length = string.length
//            return if (length > 1 && string[0] == '\"' && string[length - 1] == '\"') {
//                string.substring(1, length - 1)
//            } else string
        }
    }

    class ConnectivityStatus {
        var mMobileNetworkName: String? = null
        var mMobileSignalStrength = 0
        var mNetworkType = 0
        var mWifiSignalStrength = 0
        var mWifiSsid: String? = null
    }

    private class LeanbackLauncherPhoneStateListener(listener: ConnectivityListener?) :
        PhoneStateListener() {
        private val mListener: WeakReference<ConnectivityListener?> =
            WeakReference<ConnectivityListener?>(listener)

        override fun onSignalStrengthsChanged(signalStrength: SignalStrength) {
            super.onSignalStrengthsChanged(signalStrength)
            val listener = mListener.get()
            if (listener != null) {
                listener.connectivityStatus.mMobileSignalStrength = getLevel(signalStrength)
            }
        }

    }

    fun start() {
        if (!mIsRegistered) {
            updateConnectivityStatus()
            mContext.registerReceiver(mReceiver, mFilter)
            mTelephonyManager!!.listen(mPhoneStateListener, 256)
            mIsRegistered = true
        }
    }

    fun stop() {
        if (mIsRegistered) {
            mContext.unregisterReceiver(mReceiver)
            mTelephonyManager!!.listen(mPhoneStateListener, 0)
            mIsRegistered = false
        }
    }

    private fun setNoConnection() {
        connectivityStatus.mNetworkType = 1
        connectivityStatus.mWifiSsid = null
        connectivityStatus.mWifiSignalStrength = 0
    }

    private fun isSecureWifi(wifiInfo: WifiInfo?): Boolean {
        if (wifiInfo == null) {
            return false
        }
        val networkId = wifiInfo.networkId
        val configuredNetworks = if (ActivityCompat.checkSelfPermission(
                mContext,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return false
        } else
            mWifiManager.configuredNetworks ?: return false
        for (configuredNetwork in configuredNetworks) {
            if (configuredNetwork.networkId == networkId) {
                return configuredNetwork.allowedKeyManagement[1] || configuredNetwork.allowedKeyManagement[2] || configuredNetwork.allowedKeyManagement[3]
            }
        }
        return false
    }

    private fun updateConnectivityStatus() {
        // deprecated in API29
        val networkInfo = mConnectivityManager.activeNetworkInfo
        if (networkInfo == null) {
            setNoConnection()
            return
        }
        val isConnected = readConnectivity(mContext)
        when (networkInfo.type) {
            0 -> {
                if (isConnected) {
                    connectivityStatus.mNetworkType = 13
                } else {
                    connectivityStatus.mNetworkType = 15
                }
                var operator: String? = null
                if (mTelephonyManager != null) {
                    operator = mTelephonyManager.networkOperatorName
                    if (operator != null) {
                        operator = removeDoubleQuotes(operator)
                    }
                }
                if (operator != null) {
                    connectivityStatus.mMobileNetworkName = operator
                } else {
                    connectivityStatus.mMobileNetworkName = ""
                }
                connectivityStatus.mWifiSsid = null
                connectivityStatus.mWifiSignalStrength = 0
                return
            }
            1 -> {
                val wifiInfo = mWifiManager.connectionInfo
                if (isSecureWifi(wifiInfo)) {
                    connectivityStatus.mNetworkType = 5
                } else {
                    connectivityStatus.mNetworkType = 3
                }
                if (!isConnected) {
                    connectivityStatus.mNetworkType = 9
                }
                var ssid: String? = null
                if (wifiInfo != null) {
                    ssid = wifiInfo.ssid
                    if (ssid != null) {
                        ssid = removeDoubleQuotes(ssid)
                    }
                }
                if (ssid != null) {
                    connectivityStatus.mWifiSsid = if (ssid == "<unknown ssid>") "" else ssid
                } else {
                    connectivityStatus.mWifiSsid = ""
                }
                if (wifiInfo != null) {
                    connectivityStatus.mWifiSignalStrength =
                        WifiManager.calculateSignalLevel(wifiInfo.rssi, 5)
                } else {
                    connectivityStatus.mWifiSignalStrength = 0
                }
                connectivityStatus.mMobileNetworkName = null
                connectivityStatus.mMobileSignalStrength = 0
                return
            }
            9 -> {
                if (isConnected) {
                    connectivityStatus.mNetworkType = 7
                } else {
                    connectivityStatus.mNetworkType = 11
                }
                connectivityStatus.mWifiSsid = null
                connectivityStatus.mWifiSignalStrength = 0
                connectivityStatus.mMobileNetworkName = null
                connectivityStatus.mMobileSignalStrength = 0
                return
            }
            else -> {
                setNoConnection()
                return
            }
        }
    }
}